rust/kernel/spi.rs

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::bindings;
+use crate::c_types;
+use crate::error::{Error, Result};
+use crate::CStr;
+use alloc::boxed::Box;
+use core::pin::Pin;
+
+#[derive(Clone, Copy)]
+pub struct SpiDevice(*mut bindings::spi_device);
+
+impl SpiDevice {
+    pub fn from_ptr(dev: *mut bindings::spi_device) -> Self {
+        SpiDevice(dev)
+    }
+
+    pub fn to_ptr(&mut self) -> *mut bindings::spi_device {
+        self.0
+    }
+}
+
+pub struct DriverRegistration {
+    this_module: &'static crate::ThisModule,
+    registered: bool,
+    name: CStr<'static>,
+    probe: Option<SpiMethod>,
+    remove: Option<SpiMethod>,
+    shutdown: Option<SpiMethodVoid>,
+    spi_driver: Option<bindings::spi_driver>,
+}
+
+impl DriverRegistration {
+    fn new(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Self {
```

Comment:
Won't your driver require driver data, i.e. the private memory you usually `kzalloc()` in `probe()`, and attach to your driver using `spi_set_drvdata()`? I noticed that in your [driver](https://github.com/ks0n/mfrc522-linux/blob/master/rust-module/mfrc522.rs#L27) you've sidestepped this issue by storing the driver data as a `static`. But this will stop working once you instantiate multiple instances (devices) of the same class (driver). E.g. you could have two spi memory chips of the same type. On different SPI busses, or even on the same SPI bus, using a different chip-select.

We currently have a [PR](https://github.com/Rust-for-Linux/linux/pull/254/commits/d4552cdb02ee5ac0995d76607ea6263a102cb825#diff-e4c0dbff9f0e242cce9288eb63fbe113c08888747bc4ae07bd7e345818e1e34cR85) in-flight that's trying to solve the problem of driver data, you could help review?